### ISSUE TITLE
Enhancement: [no-floating-promises] Add option to ignore catches

### ISSUE THREAD

**Issue by rnveach** [typescript-eslint/typescript-eslint#11421]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises/

### Description

I propose that the rule should allow users to configure an option to ignore catches in promises and allow them to still be flagged as "floating". While this lint documentation specifically talks about a promise without error handling, I think of a "floating" promise as one having no direct connection to the rest of your code such that there is no flow control of when it interacts (or reconnects in interaction) with the rest of the code. It is "floating" in a separate space then the rest of the execution.

Take this code for example where you may try to test a method which has a promise with a catch:
```ts
  myField: number = null;

  myMethod() {
    Promise.resolve(() => {
      this.myField = 2;
    })
      .catch((error: unknown) => console.log(error));
  }

  it('should create', () => {
    component.myMethod();
    expect(component.myField).toBe(2); // field is still null, so test fails
  });
```

Adding all promise related lints with the defaults pass on this code. 
```
            "@typescript-eslint/no-deprecated": "error",
            "@typescript-eslint/require-await": "error",
            "@typescript-eslint/return-await": "error",
            "@typescript-eslint/no-misused-promises": "error",
            "@typescript-eslint/no-floating-promises": "error",
            "@typescript-eslint/use-unknown-in-catch-callback-variable": "error",
            "@typescript-eslint/prefer-promise-reject-errors": "error",
            "@typescript-eslint/promise-function-async": "error",
```

The test fails because the promise hasn't resolved yet and we can't tell when it will finish since it isn't returned or awaited on, which would fix this. If it wasn't for the catch, no-floating-promises would flag it as a violation. So the catch doesn't help it interconnect with the rest of the execution. This is the type of code I want to violate on with proposed new option.

My recommendation for a option name would be `ignoreCatches` and it would ignore `.catch(...)` and `.then()` with two arguments. It would be a boolean, and the default could be `false` to maintain backwards compatibility.

In my project, I want to be able to force chain (sequential/parallel) all promises, and connect them together to ensure we have a strict flow of execution of the promises. This would require us to always return the promises up the chain if they ever exist. This is the reason I have created this issue and requesting this option.

### Fail

```TypeScript
myMethod() {
    Promise.resolve(() => { // violation, Promises must be awaited or returned
      this.myField = 2;
    })
      .catch((error: unknown) => console.log(error));
  }
```

### Pass

```TypeScript
async myMethod() { // subsequent change from other lints
    return Promise.resolve(() => { // resolved
      this.myField = 2;
    })
      .catch((error: unknown) => console.log(error));
  }
```

**Comment by kirkwaiblinger** [comment]
Hi @rnveach!

The way I see the situation is that there are two possible underlying purposes for having the rule `no-floating-promises` (you've more or less articulated the same thing in the issue report):

1. Avoid unhandled promise rejections
1. Avoid unexpected control flow due to not awaiting async things

Obviously, these have a lot of overlap! But, like you say, the rule's current behavior around `promise.catch(<...>)` fulfills purpose 1 and does _not_ necessarily fulfill purpose 2, since we may still need to wait for the `promise` to resolve before proceeding for control flow reasons.

My question would be — can you just ban `.then()`/`.catch()`/`.finally()` completely in your project? Or are there any other edge cases you've encountered where purposes 1 and 2 come into conflict? (I can't think of any but genuinely curious)

My thought is I would much rather read
```ts
async function foo() {
  try {
    await somethingAsync();
  } catch (error: unknown) {
    console.log(error)
  }
}
```
anyway than the promise handler version. Technically this also is more robust in case `somethingAsync()` throws synchronously, rather than returning a rejected promise.

I'm _personally_ also in favor of the "purpose 2" version of this rule, but, even so, my thought has been that it doesn't warrant an option if the only difference is when using the anyways-discouraged `.then()`/`.catch()`/`.finally()` syntax.

Thoughts?

**Comment by kirkwaiblinger** [comment]
Also, real quick, is the example code "real" code or example code? Because, the example doesn't ever execute the `this.myField = 2` line... and wouldn't regardless of the `ignoreCatch`. 

`Promise.resolve(() => { console.log('I am just a function') })` evaluates to a promise whose fulfillment value is the arrow function, but it never executes the function. You would have to use an IIFE or something like that...

```ts
myMethod() {
  (async () => {
    this.myField = 2;
  })().catch((error: unknown) => console.log(error));
}
```

Just asking because I'm wondering whether you've encountered a real bug that this proposed check would fix (and made a trivial mistake when paring the code down for demoing) or the bug you encountered is actually slightly different...

**Comment by rnveach** [comment]
>  a real bug

I am not sure I would go as far as to declare this a "bug". I don't have any concrete cases of we had an issue in production, and it was connected to this promise code. This issue mostly stems from I see something missing with all these lints, and would like to see a way to be able to control promises a bit more to narrow future issues.

> Just asking because I'm wondering whether you've encountered a real bug that this proposed check would fix (and made a trivial mistake when paring the code down for demoing) or the bug you encountered is actually slightly different...

This is most likely my fault from trying to demo it down.

Most of the code I see in my project is like:
````ts
import { lastValueFrom } from 'rxjs';

  myMethod() {
    lastValueFrom(this.service.getData()).then(data => {
      processAndAssignData(data);
    })
      .catch((error: unknown) => handleError(error));
  }
````

All of our service calls have some sort of console logging and error handling attached to them to assist to identify the failures and send users to an error page.

> My question would be — can you just ban .then()/.catch()/.finally() completely in your project? Or are there any other edge cases you've encountered where purposes 1 and 2 come into conflict? (I can't think of any but genuinely curious)

Unless there is a lint which specifically controls it, I don't see a way to fully ban something in my project. Communication and coordination is low among developers outside of inter-dependencies. I also understand lint can't cover every scenario, but I thought this could be covered here since catches are already identified.

It is more common we use `catch` than `finally`, so I didn't really have concerns or even think of it. Even still, I don't see banning catches as a good thing. We use them to log errors and send the users to error pages. It makes sense to identify which service broke incase we have to debug an issue. We don't yet have a global error handler.

My intention wasn't to ban `then` as I want to chain promises together as I mentioned above. In my example `myMethod` above, the newly returned promise could be chained using `then` for the next promise or the rest of the execution. It would still be up to the developer to chain (then) or parallel (Promise.all).

> My thought is I would much rather read
> try { .. await  .. } catch { .. }

My plan was to avoid using `await` as much as possible, in such to ban it using `no-restricted-syntax`. While the code may be easier to read, `await` does more to prevent parallel promises and I feel our project likes to overuse this. For the most part we just await on every promise created. There has been a complaint that our pages take too long to load. This is what prompted to me to look at all the promise lints.

**Comment by kirkwaiblinger** [comment]
> > a real bug
> 
> I am not sure I would go as far as to declare this a "bug". I don't have any concrete cases of we had an issue in production, and it was connected to this promise code. This issue mostly stems from I see something missing with all these lints, and would like to see a way to be able to control promises a bit more to narrow future issues.

Understood 👍 

> Unless there is a lint which specifically controls it, I don't see a way to fully ban something in my project. Communication and coordination is low among developers outside of inter-dependencies. I also understand lint can't cover every scenario, but I thought this could be covered here since catches are already identified.

Here's a basic version with no-restricted-syntax ([playground](https://typescript-eslint.io/play/#ts=5.8.2&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPACjEgLYCWaIAPCgK7EBGIMAfANwBQUAdBgBYgoAFAEoO3MFAxheIsVwBmpFFAgQAnrPaceSAMoYYSgOabuGJABkkEiCH2GUJ0UA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1IB3NtFyQANJnDYcyREhr5o3SAFlEAWwBGiaAFEOABxLJkFfLjRq2pKgAsUACh6H8%2B7aSYA6XGzWIAvADkpHa4AQC64k7QLm6e3r6BVFa24ZHOrtDuXj7%2BAQBmFN7w8EzhAJRiklKQvsZsAOaISnRmQWCwMmDOahSdNmy4dEjQqFVgAL5hkuMg40A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

> It is more common we use `catch` than `finally`, so I didn't really have concerns or even think of it. Even still, I don't see banning catches as a good thing. We use them to log errors and send the users to error pages. It makes sense to identify which service broke incase we have to debug an issue. We don't yet have a global error handler.
> 
> My intention wasn't to ban `then` as I want to chain promises together as I mentioned above. In my example `myMethod` above, the newly returned promise could be chained using `then` for the next promise or the rest of the execution. It would still be up to the developer to chain (then) or parallel (Promise.all).
> 
> > My thought is I would much rather read
> > try { .. await  .. } catch { .. }
> 
> My plan was to avoid using `await` as much as possible, in such to ban it using `no-restricted-syntax`. While the code may be easier to read, `await` does more to prevent parallel promises and I feel our project likes to overuse this. For the most part we just await on every promise created. There has been a complaint that our pages take too long to load. This is what prompted to me to look at all the promise lints.

Hmm, yeah, it might look a little wonky at first glance, but I personally use Async IIFEs for this. 

```ts
async function doAsyncStuffParallel() {
  const promise1 = (async () => {
    await 'something';
    await 'something else';
  })();

  const promise2 = (async () => {
    await 'still something else';
    try {
      await 'something that might throw!'
    } catch (e) {
      // handle error
    }
  })();

  await Promise.all([promise1, promise2]);
}
```

**Comment by JoshuaKGoldberg** [comment]
Indeed, the rule is working as expected. There've been a lot of attempts to make `catch (error) { ` / `catch(error => ...)`s type-safe and lintable with NFP over the years. None have resulted in proposals solid enough to add here. But if someone has any concrete ideas, that'd be great, please do post 🙂 .

This was a good discussion. Closing out as it's been a while without new info. Thanks & cheers!
